### ISSUE TITLE
Which (web) UI can be used with MemGPT?

### ISSUE THREAD

**Issue by distributev** [cpacker/MemGPT#692]
Question: Does MemGPT has its own web user interface or it can be used with existing web interfaces like

https://github.com/oobabooga/text-generation-webui

or 

https://github.com/mckaywrigley/chatbot-ui

?

Does MemGPT mimics ChatGPT public API (so that it can be used with existing web UIs mentioned above)?

It is not very clear to me how MemGPT is supposed to be used by normal users much like ChatGPT is used by so many users using the default ChatGPT user interface. Is this something which you never intended to achieve with MemGPT?

I mean I saw that MemGPT has a command line from where it can be used and I also know that it can be integrated with Autogen and other tools in a (API?) kind of way but what about (in addition to the existing command line and API?) a web user interface?

**Comment by cpacker** [comment]
tl;dr we are developing a MemGPT-native Chat UI that will be released as part of the MemGPT repo / package, this is coming ASAP and you can track progress here: https://github.com/cpacker/MemGPT/pull/627

MemGPT is *stateful*, so you can't use it via a traditional ChatCompletion API which is stateless, so it can't easily be integrated into existing chat UI frameworks / frontends that are designed around the ChatCompletion API.

MemGPT is much closer to the new Assistants API, so once existing frontends add support for that (neither Web UI nor ChatBot UI do AFAIK) it will be much more straightforward to plug MemGPT into those UI frameworks. MemGPT actually came before the Assistants API, but we will be aligning our API to mirror the OpenAI Assistants API for easy compatibility and development.

Additionally, MemGPT has some specific design features like memory tiers (core memory, archival memory) and persona files (human and bot/persona files) which warrant their own UI components in a chat UI. We would either have to build these into an existing chat UI (on top of adding support for stateless/Assistants-style API), or we can build our own light chat UI that sits on top of MemGPT. The former is much more work than the latter, so we're starting by rolling our own chat UI which will be available soon (see the PR or our Discord for updates).

**Comment by distributev** [comment]
Another option would be to provide sample/boilerplate code how existing libraries like streamlit, gradio, chainlit, etc can be used to build custom UIs on top of MemGTP. 

I enumerated these libraries as these are existing libs which people use to (easily/fast) build custom UIs on top of LLMs (less effort involved vs building the same UI using a full blow framework like Flask)

I'm just learning about MemGPT so for the moment I do not know enough to understand if the stateful/stateless limitation apply to these libs as well (or it does not apply since, being libraries, one can build their custom UI having more state control vs pre-built UIs like web UI or ChatBot UI)

https://github.com/streamlit/streamlit
https://github.com/gradio-app/gradio
https://github.com/Chainlit/chainlit

https://github.com/antoineross/Autogen-UI - Chainlit on top of Autogen

**Comment by cpacker** [comment]
> I enumerated these libraries as these are existing libs which people use to (easily/fast) build custom UIs on top of LLMs (less effort involved vs building the same UI using a full blow framework like Flask)

Yeah I definitely agree with the general idea of building off of as much OSS components as possible - just to give you some insight, we initially attempted to build a MemGPT UI backwards off of an existing ChatBot UI style repo and immediately realized that there was a significant amount of code that needed to get rewritten to account for the assumptions made about ChatCompletion (in how the UI worked / handled state), and ultimately it would be faster for us to go in the other direction and build a simple chat app using premade components on top of our own Assistants-style API (WIP: https://memgpt.readme.io/reference/api).

I also 100% agree that it would be amazing to have some boilerplate code to show people how to integrate MemGPT into popular libraries like gradio. If this is something you might be interested in contributing as a PR, we'd be happy to help you! We might eventually get to it ourselves, but I think this would also make a great community PR.

**Comment by distributev** [comment]
Yes, I would be interested. How we'd communicate further?

**Comment by cpacker** [comment]
Awesome - the best way is probably to join the MemGPt Discord server (link on the front README and at memgpt.ai, and via that server you can them DM me (@cpacker).

If you don't use Discord, the best way is to probably open a GitHub issue to track your planned feature, then eventually open a draft PR. You should still open a GitHub issue and/or PR but if you're also on Discord it's easier to communicate over there, if not we can chat via github issue / PR comments.

**Comment by distributev** [comment]
I've contacted you on Discord.

**Comment by cpacker** [comment]
For future reference, a big (community) PR by @BabellDev was merged that adds a clean way to interact with MemGPT via python (via a "python client"): #713. This should be very useful for connecting to existing Python chat/web UI interfaces like gradio.
